File: python/utils/encryptions.py
import base64
import hashlib
from cryptography.hazmat.primitives.ciphers.aead import AESGCM

def decrypt_text(payload: str, password: str) -> str:
    data = base64.b64decode(payload)

    salt = data[0:16]
    iv = data[16:28]
    auth_tag = data[28:44]
    ciphertext = data[44:]

    # Node's crypto.scryptSync defaults: N=16384, r=8, p=1, keylen=32
    key = hashlib.scrypt(
        password.encode("utf-8"),
        salt=salt,
        n=16384,
        r=8,
        p=1,
        dklen=32,
    )

    aesgcm = AESGCM(key)
    # Python's AESGCM expects ciphertext + tag concatenated
    plaintext = aesgcm.decrypt(iv, ciphertext + auth_tag, None)
    return plaintext.decode("utf-8")

File: src/firewall/encryptions.ts
import crypto from "crypto";

const ENCRYPTION_ALGO = "aes-256-gcm";

function deriveKey(password: string, salt: Buffer): Buffer {
    // scrypt derives a fixed-length key from your password; salt keeps it unique per record
    return crypto.scryptSync(password, salt, 32);
}

export const encryptText = (plaintext: string, password: string): string => {
    const salt = crypto.randomBytes(16);
    const iv = crypto.randomBytes(12); // 12 bytes is standard for GCM
    const key = deriveKey(password, salt);

    const cipher = crypto.createCipheriv(ENCRYPTION_ALGO, key, iv);
    const encrypted = Buffer.concat([cipher.update(plaintext, "utf8"), cipher.final()]);
    const authTag = cipher.getAuthTag();

    // pack salt + iv + authTag + ciphertext together so decrypt has everything it needs
    return Buffer.concat([salt, iv, authTag, encrypted]).toString("base64");
};

export const decryptText = (payload: string, password: string): string => {
    const data = Buffer.from(payload, "base64");
    const salt = data.subarray(0, 16);
    const iv = data.subarray(16, 28);
    const authTag = data.subarray(28, 44);
    const encrypted = data.subarray(44);
    const key = deriveKey(password, salt);

    const decipher = crypto.createDecipheriv(ENCRYPTION_ALGO, key, iv);
    decipher.setAuthTag(authTag);
    const decrypted = Buffer.concat([decipher.update(encrypted), decipher.final()]);
    return decrypted.toString("utf8");
};